test-lib.sh

```diff
@@ -765,6 +765,83 @@ EOF
     )
 }
 
+# ct_s2i_multistage_build APP_PATH SRC_IMAGE DST_IMAGE SEC_IMAGE [S2I_ARGS]
+# ----------------------------
+# Create a new s2i app image from local sources in a similar way as source-to-image would have used.
+# Argument: APP_PATH - local path to the app sources to be used in the test
+# Argument: SRC_IMAGE - image to be used as a base for the s2i build process
+# Argument: SEC_IMAGE - image to be used as the base for the result of the build process
+# Argument: DST_IMAGE - image name to be used during the tagging of the s2i build result
+# Argument: S2I_ARGS - Additional list of source-to-image arguments.
+#                      Only used to check for environment variable definitions.
+ct_s2i_multistage_build() {
+
+  local app_path=$1; shift
+  local src_image=$1; shift
+  local sec_image=$1; shift
+  local dst_image=$1; shift
+  local s2i_args=$*;
+  local local_app="app-src"
+  local user_id=
+  local mount_options=
+
+
+  # Run the entire thing inside a subshell so that we do not leak shell options outside of the function
+  (
+  # Error out if any part of the build fails
+  set -e
+
+  user=$(docker inspect -f "{{.Config.User}}" "$src_image")
+  # Default to root if no user is set by the image
+  user=${user:-0}
+  # run the user through the image in case it is non-numeric or does not exist
+  # NOTE: The '-eq' test is used to check if $user is numeric as it will fail if $user is not an integer
+  if ! [ "$user" -eq "$user" ] 2>/dev/null && ! user_id=$(docker run --rm "$src_image" bash -c "id -u $user 2>/dev/null"); then
+      echo "ERROR: id of user $user not found inside image $src_image."
+      echo "Terminating s2i build."
+      return 1
+  else
+      user_id=${user_id:-$user}
+  fi
+
+  # Use /tmp to not pollute cwd
+  tmpdir=$(mktemp -d)
+  df_name=$(mktemp -p "$tmpdir" Dockerfile.XXXX)
+  cd "$tmpdir"
+
+  # Strip file:// from APP_PATH and copy its contents into current context
+  mkdir -p "$local_app"
+  cp -r "${app_path/file:\/\//}/." "$local_app"
+
+  cat <<EOF >"$df_name"
```

Comment:
Is it a good idea to hardcode the Dockerfile here? With all I learned during preps of Python minimal container, I can imagine that Dockerfile can be provided to this function and just slightly updated as `ct_test_app_dockerfile` does now. Mainly because minimal containers usually need to install some additional packages and do some other hack and there is no way how to do it now with hardcoded dockerfile.

Reply:
We can add an argument for passing an existing Dockerfile and modifying it as necessary